### ISSUE TITLE
Support Material 3 motion - Transitions

### ISSUE THREAD

**Issue by guidezpl** [flutter/flutter#116526]
https://m3.material.io/styles/motion/transitions/transition-patterns

There are now six common transition patterns:
- Container transform
- Forward and backward
- Lateral
- Top level
- Enter and exit
- Skeleton loaders

This work includes:
- [ ] making all of these available and to spec in the [animations](https://pub.dev/packages/animations) package
- [ ] ensuring framework transitions use these transitions rather than hardcoded defaults

**Comment by florian-obernberger** [comment]
Will this also include the page transition mentioned in #110664?

And does anyone know where the specifications for this transition are?

**Comment by guidezpl** [comment]
Yes, but that'll be in framework rather than the `animations` package. Re the spec, I don't know

**Comment by masreplay** [comment]
What about the skeleton loaders?

**Comment by guidezpl** [comment]
> What about the skeleton loaders?

There are many packages on pub that do support this. Since the guidelines don't specify a specific kind of loader, I don't think we'll add support for them in the framework yet.

**Comment by HansMuller** [comment]
CC @justinmc

**Comment by justinmc** [comment]
It looks like some of these transitions are also affected by the predictive back feature.  I'm going to make sure we have a good way to respond to those back gestures and will keep in mind they need to support these kinds of animations.

| | Transitions | Predictive Back |
| --- | --- | --- |
| Video | <video src="https://user-images.githubusercontent.com/389558/217968018-84ae865a-101d-45bb-b159-2c3a258f9f38.mp4" /> | <video src="https://user-images.githubusercontent.com/389558/217968398-45bcdb7d-a932-449c-82c3-aa6422c148db.mp4" /> |
| Spec | https://m3.material.io/styles/motion/transitions/transition-patterns#8c2796f3-6596-4709-83b8-061c78ffd55d | https://carbon.googleplex.com/predictive-back-for-apps/pages/at-2-floating-surfaces |

**Comment by dhushyanth-s** [comment]
I think this issue should be updated to reflect the fact that some of the transitions listed above are already available in the animations package and are ready to use. I was searching for a way to use Material 3 transitions in Flutter earlier this week and this issue came up pretty high on search results but was really misleading byt making my think they were not available at all in flutter.

**Comment by GCSBOSS** [comment]
Is this planned for the near future? A lot of Material Design value comes from motion.

**Comment by guidezpl** [comment]
I added the curves and durations with https://github.com/flutter/flutter/pull/129942, so this issue is unblocked for any contributor.

**Comment by tbm98** [comment]
Will these features be implemented, while the main issue #91605 is closed?

**Comment by crystalarcus** [comment]
> the transitions listed above are already available in the animations package and are ready to use.

@dhushyanth-s 
Yes and No.
Yes, most transitions are available in `animations` package although, they are from previous material design and not Material 3. There is a lot of difference in some things, especially in easing and duration. 
For example, the equivalent for Container Transform in the animation package is  `OpenContainer()`,
which can be used for transition between 2 Widgets. But it takes only 1 argument for the duration, which it uses for both opening and closing animation, while Material 3 suggests using a longer duration while opening and shorter for closing.
I tried to customize it with `setState`, and give duration as a variable, but still, the duration is the same while opening and closing. 
Guess we need a new package with Material 3 Specifications.

**Comment by guidezpl** [comment]
> Guess we need a new package with Material 3 Specifications.

Contributions to `animations` are welcome

**Comment by justinmc** [comment]
I'm working on the OpenContainer transition's predictive back support here: https://github.com/flutter/packages/pull/6321

I might mess with trying to match the M3 spec better if it's not a big addition to the PR, but people should feel free to open their own PR tweaking that in the meantime.

**Comment by crystalarcus** [comment]
> I'm working on the OpenContainer transition's predictive back support here: [flutter/packages#6321](https://github.com/flutter/packages/pull/6321)

Yesterday, I edited `OpenContainer` from  `animations` package and added M3's `easeInOutCubicEmphasized` with 500ms duration when opening Container and `Easing.EmphasizedDecelerator` with 200ms duration when closing Container, and it works just find, like M3 Container transform.

**Comment by crystalarcus** [comment]
> Yesterday, I edited `OpenContainer` from `animations` package and added M3's `easeInOutCubicEmphasized` with 500ms duration when opening Container and `Easing.EmphasizedDecelerator` with 200ms duration when closing Container, and it works just find, like M3 Container transform.
>
Here's a video it, it's my project Instagram Redesigned with Material 3:

https://github.com/flutter/flutter/assets/150055488/6857882b-7452-428a-bda3-bb17264f6d2b

it's seems a little laggy in video, due to low fps, but it works smooth in app.

**Comment by swch01** [comment]
Where can I find detailed specs on the transitions? The Material 3 site defines duration and easing values but does not go into detail about the transitions. 

For example, regarding the Top Level transition, the site says: "The exiting screen quickly fades out and then the entering screen fades in." The "Guidelines" link does not provide details, and the "Android implementation" link provides Material 2 details.

All I have are the examples. When I step through them frame by frame, I see the exiting screen fading to about 50% opacity before being replaced with the entering screen at about 50% opacity.

I'm able to implement the transitions well enough in this manner, but I would appreciate a look at a spec.

**Comment by l619534951** [comment]
> Where can I find detailed specs on the transitions? The Material 3 site defines duration and easing values but does not go into detail about the transitions.

As a non-designer, I am also confused when making my own app, so I am using mdc-android animation duration. 
By the way, mdc-android, jetpack compose,  flutter default material animation, color are a little bit not the same.
For example, a Scaffold that contains both AppBar and ListView, and when the ListView scrolls, their AppBar colors are different

**Comment by crystalarcus** [comment]
> Where can I find detailed specs on the transitions? The Material 3 site defines duration and easing values but does not go into detail about the transitions.

@swch01 You will have to use one of defined easings and durations, according to your need. Normally, as said in Material Design Guidelines, use emphasizedDeccelerator when entering screen, and emphasizedAccelerator when exiting screen, and for duration, entering Duration > exiting duration, for example u can keep, entering Duration : 400ms and exiting Duration as 300ms

**Comment by guidezpl** [comment]
FYI M3 curves are in https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/motion.dart, please use those instead of creating duplicates

**Comment by ercantomac** [comment]
Any updates on this?

**Comment by brainer3220** [comment]
> Any updates on this?

Probably no.

**Comment by spauldhaliwal** [comment]
> FYI M3 curves are in https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/motion.dart, please use those instead of creating duplicates

This is missing the Emphasized curve, which is the first one described [here](https://m3.material.io/styles/motion/easing-and-duration/tokens-specs), and has coordinate values of x0:0.2, y0:0, x1:0, y1:1

**Comment by xVemu** [comment]
> This is missing the Emphasized curve

It's located in [Curves.easeInOutCubicEmphasized](https://api.flutter.dev/flutter/animation/Curves/easeInOutCubicEmphasized-constant.html)

**Comment by Sesa1988** [comment]
As it stands, it seems you can't directly use the animations package with named routes. It would be more flexible if we could declare multiple PageTransitionsThemes for a single platform and then, when pushing new pages, specify which transition to use, perhaps by passing a key or identifier. While it's likely more complex than that, having such an option would make it easier to manage different animations for different routes.

**Comment by IanVS** [comment]
Hi, I'm a pretty new flutter dev, looking for a way to make `Navigator.push()` use a Material 3 forward/back animation.  Is this currently not possible, given that this issue is still open?  The interactive example at https://docs.flutter.dev/cookbook/navigation/navigation-basics#interactive-example makes it seem like it works by default, but I can't replicate that in my Android app.

**Comment by justinmc** [comment]
@IanVS Maybe you're looking for the transition added in https://github.com/flutter/flutter/pull/158881, which is currently on master but not beta or stable?

**Comment by IanVS** [comment]
😍 that looks like exactly what I need! Is there any way to know when that will be released?

**Comment by justinmc** [comment]
Looks like it's on beta and will be in the next stable release. https://justinmc.github.io/flutter_releases/#/pr/framework/158881

**Comment by IanVS** [comment]
Oh whoa, I was asking the other day in discord how to know whether a PR was in a beta or not, this seems like a really cool tool, thanks!

**Comment by justinmc** [comment]
Disclaimer that it often hits GitHub's rate limiting because it's a major hack haha.
